File: leetcode/medium/49-groupAnagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> output = new ArrayList<List<String>>();
        HashMap<String, List<String>> hashMap = new HashMap<String, List<String>>();
        
        for (String s : strs) {
            String[] arr = s.split("");
            Arrays.sort(arr);
            String curr = String.join("", arr);
            
            if (hashMap.get(curr) != null) {
                hashMap.get(curr).add(s);
            } else {
                hashMap.put(curr, new ArrayList<String>());
                hashMap.get(curr).add(s);
            }
        }
        
        for (List<String> arr : hashMap.values()) {
            output.add(arr);
        }
        return output;
    }
}

File: leetcode/medium/49-groupAnagrams.js
// Given an array of strings strs, group the anagrams together. You can return the answer in any order.

// An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all 
// the original letters exactly once.

// Example 1:

// Input: strs = ["eat","tea","tan","ate","nat","bat"]
// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
// Example 2:

// Input: strs = [""]
// Output: [[""]]
// Example 3:

// Input: strs = ["a"]
// Output: [["a"]]

// first pass solution
var groupAnagrams = function(strs) {
    let dict = {};
    
    for (let str of strs) {
        let temp = str.split("").sort().join("");
        if (!(temp in dict)) {
            dict[temp] = [str];
        } else {
            dict[temp].push(str);
        }
    }
    return Object.values(dict);
};
